src/lib/provable/gadgets/sha2.ts

```diff
@@ -0,0 +1,621 @@
+  SIGMA_ONE: {
+    // SHA2-224 and SHA2-256: §4.1.2 eq.4.5
+    224: [6, 11, 25] as TupleN<number, 3>,
+    256: [6, 11, 25] as TupleN<number, 3>,
+    // SHA2-384 and SHA2-512: §4.1.3 eq.4.11
+    384: [14, 18, 41] as TupleN<number, 3>,
+    512: [14, 18, 41] as TupleN<number, 3>,
+  },
+
+  // Initia hash values
+  H: {
+    // SHA2-224 §5.3.2
+    224: [
+      0xc1059ed8, 0x367cd507, 0x3070dd17, 0xf70e5939, 0xffc00b31, 0x68581511,
+      0x64f98fa7, 0xbefa4fa4,
+    ],
+    // SHA-256 §5.3.3
+    256: [
+      0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c,
+      0x1f83d9ab, 0x5be0cd19,
+    ],
+    // SHA2-384 §5.3.4
+    384: [
+      0xcbbb9d5dc1059ed8n,
+      0x629a292a367cd507n,
+      0x9159015a3070dd17n,
+      0x152fecd8f70e5939n,
+      0x67332667ffc00b31n,
+      0x8eb44a8768581511n,
+      0xdb0c2e0d64f98fa7n,
+      0x47b5481dbefa4fa4n,
+    ],
+    // SHA2-512 §5.3.5
+    512: [
+      0x6a09e667f3bcc908n,
+      0xbb67ae8584caa73bn,
+      0x3c6ef372fe94f82bn,
+      0xa54ff53a5f1d36f1n,
+      0x510e527fade682d1n,
+      0x9b05688c2b3e6c1fn,
+      0x1f83d9abfb41bd6bn,
+      0x5be0cd19137e2179n,
+    ],
+  },
+
+  K: {
+    224: K224_256,
+    256: K224_256,
+    384: K384_512,
+    512: K384_512,
+  },
+};
+
+const SHA2 = {
+  /**
+   * Implementation of [NIST SHA-2](https://csrc.nist.gov/pubs/fips/180-4/upd1/final)
+   * hash Function. Supports output lengths of 224, 256, 384, or 512 bits.
+   *
+   * Applies the SHA-2 hash function to a list of big-endian byte-sized {@link Field}
+   * elements, flexible to handle varying output lengths (224, 256, 384, 512 bits) as specified.
+   *
+   * The function accepts {@link Bytes} as the input message, which is a type that
+   * represents a static-length list of byte-sized field elements (range-checked
+   * using {@link Gadgets.rangeCheck8}).
+   * Alternatively, you can pass plain `number[]` of `Uint8Array` to perform a hash
+   * outside provable code.
+   *
+   * Produces an output of {@link Bytes} that conforms to the chosen bit length.
+   * Both input and output bytes are big-endian.
+   *
+   * @param len - Desired output length in bits. Valid options: 224, 256, 384, 512.
+   * @param message - Big-endian {@link Bytes} representing the message to hash.
+   *
+   * ```ts
+   * let preimage = Bytes.fromString("hello world");
+   * let digest224 = SHA2.hash(224, preimage);
+   * let digest256 = SHA2.hash(256, preimage);
+   * let digest384 = SHA2.hash(384, preimage);
+   * let digest512 = SHA2.hash(512, preimage);
+   * ```
+   *
+   */
+  hash<T extends Length>(length: T, data: FlexibleBytes): Bytes {
+    // Infer the type T based on the value of `length` (conditional type)
+    type Type = T extends 224 | 256 ? UInt32 : UInt64;
+
+    // preprocessing §6.2
+    // padding the message $5.1.1 into blocks that are a multiple of 512
+    let messageBlocks = padding<Type>(length, data);
+
+    let H = SHA2.initialState<Type>(length);
```

Comment:
Nit: it seems we're doing the Uint32/Uint64 transformations every time we execute `hash`. A direct optimization would be to do it once and for all.

Not blocking.